File: MathParserLib/Function.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MathParserLib
    {
    public abstract class Function
        {
        protected List<Value> args = new List<Value>();
        public abstract string getName();
        public void clearArgs() { args.Clear(); }
        public void addArg(Value arg) => args.Add(arg);
        public abstract Value execute();

        private static Dictionary<string, Function> functions;

        static void AddFunction(Function f)
            {
            functions.Add(f.getName(), f);
            }
        static Function()
            {
            functions = new Dictionary<string, Function>();
            AddFunction(new Sin());
            AddFunction(new Cos());
            AddFunction(new Tan());
            AddFunction(new ArcSin());
            AddFunction(new ArcCos());
            AddFunction(new ArcTan());
            AddFunction(new Asin());
            AddFunction(new Acos());
            AddFunction(new Atan());
            AddFunction(new Sqrt());
            }

        public static Function get(string functionName)
            {
            if (functions.ContainsKey(functionName))
                return functions[functionName];
            else
                return null;
            }
        public static bool exists(string functionName) => functions.ContainsKey(functionName);
        }


    public class Sin : Function
        {
        public override string getName() => "sin";
        public override Value execute()
            {
            if (args.Count != 1)
                return Double.NaN;
            double arg = args[0].number;
            if (args[0].unit != null && args[0].unit.Type == TokenType.DEG)
                arg = arg * Math.PI / 180;
            return Math.Sin(arg);
            }
        }

    public class Cos : Function
        {
        public override string getName() => "cos";
        public override Value execute()
            {
            if (args.Count != 1)
                return Double.NaN;
            double arg = args[0].number;
            if (args[0].unit != null && args[0].unit.Type == TokenType.DEG)
                arg = arg * Math.PI / 180;
            return Math.Cos(arg);
            }
        }

    public class Tan : Function
        {
        public override string getName() => "tan";
        public override Value execute()
            {
            if (args.Count != 1)
                return Double.NaN;
            double arg = args[0].number;
            if (args[0].unit != null && args[0].unit.Type == TokenType.DEG)
                arg = arg * Math.PI / 180;
            return Math.Tan(arg);
            }
        }

    public class ArcTan : Function
        {
        public override string getName() => "arctan";
        public override Value execute()
            {
            if (args.Count != 1)
                return Double.NaN;

            return Math.Atan(args[0].number);
            }
        }
    public class Atan : ArcTan { public override string getName() => "atan"; }

    public class ArcSin : Function
        {
        public override string getName() => "arcsin";
        public override Value execute()
            {
            if (args.Count != 1)
                return Double.NaN;

            return new Value(Math.Asin(args[0].number), new Token(TokenType.RAD));
            }
        }
    public class Asin : ArcSin { public override string getName() => "asin"; }

    public class ArcCos : Function
        {
        public override string getName() => "arccos";
        public override Value execute()
            {
            if (args.Count != 1)
                return Double.NaN;

            return new Value(Math.Acos(args[0].number), new Token(TokenType.RAD));
            }
        }
    public class Acos : ArcCos { public override string getName() => "acos"; }

    public class Sqrt : Function
        {
        public override string getName() => "sqrt";
        public override Value execute()
            {
            if (args.Count != 1)
                return Double.NaN;

            return Math.Sqrt(args[0].number);
            }
        }
    }

File: MathParserDll/Function.cpp
#include "pch.hpp"
#include "Function.h"
#include "Scope.h"
#include "Globals.h"
#include "Tokenizer.h"
#include "Resolver.h"
#include "Date.h"
#include "GlobalFunctions.h"
#include "FunctionsView.h"

void FunctionDefs::init()
    {
    Add(new Now());
    Add(new Abs());
    Add(new Max());
    Add(new Min());
    Add(new Sin());
    Add(new Cos());
    Add(new Tan());
    Add(new ArcSin());
    Add(new ArcCos());
    Add(new ArcTan());
    Add(new ASin());
    Add(new ACos());
    Add(new ATan());
    Add(new Sqrt());
    Add(new Inc());
    Add(new Dec());
    Add(new Round());
    Add(new Floor());
    Add(new Ceil());
    Add(new Trunc());
    Add(new Factors());
    Add(new DateFunc());
    Add(new Sort());
    Add(new Reverse());
    Add(new Factorial());
    };

FunctionDef::FunctionDef(const std::string& name, size_t minArgs, size_t maxArgs)
    : name(name), minArgs(minArgs), maxArgs(maxArgs)
    {}

bool FunctionDef::isCorrectArgCount(size_t argCnt)
    {
    return argCnt >= minArgs && argCnt <= maxArgs;
    }

Value FunctionDef::call(Scope& scope, std::vector<Value>& args, const Range& range)
    { 
    if (!isCorrectArgCount(args.size()))
        return Value(Number(std::numeric_limits<double>::quiet_NaN(), 0, range));

    return execute(scope, args, range); 
    }

bool FunctionDefs::exists(const std::string& functionName)
    {
    return functions.count(functionName) != 0;
    }

void FunctionDefs::Add(FunctionDef* f) { functions.emplace(f->name, f); }

FunctionDef* FunctionDefs::get(const std::string& name)
    {
    if (functions.count(name) == 0)
        return nullptr;
    return functions.find(name)->second;
    }

CustomFunction::CustomFunction(FunctionDefExpr& functionDefExpr, CodeBlock&& codeBlock)
    : FunctionDef(functionDefExpr.id, functionDefExpr.params.size(), functionDefExpr.params.size()), functionDefExpr(functionDefExpr), codeBlock(std::move(codeBlock))
    {
    }


Value CustomFunction::execute(Scope& scope, std::vector<Value>& args, const Range& range)
    {
    std::map<std::string, Value> paramVariables;

    //args.size and params.size should be equal.
    for (size_t i = 0; i < args.size(); i++)
        {
        paramVariables.emplace(functionDefExpr.params[i], args[i]);
        }
    codeBlock.scope->setVariables(paramVariables);
    Resolver resolver(codeBlock);
    return resolver.resolveFunctionBlock(range,functionDefExpr.id);
    }

std::vector<Error> CustomFunction::getErrors()
    {
    std::vector<Error> errors;
    for (auto& stmt : codeBlock.statements)
        {    
        if (stmt->error.id != ErrorId::NONE)
            errors.push_back(stmt->error);
        }
    if(!functionDefExpr.error.isNone())
        errors.push_back(functionDefExpr.error);
    return errors;
    }

std::vector<std::string> FunctionKeys::trig = { "sin", "cos", "tan", "asin", "acos", "atan", "arcsin", "arccos", "arctan"};
std::vector<std::string> FunctionKeys::arithm = { "round", "trunc", "floor", "ceil", "abs", "max", "min"};
std::vector<std::string> FunctionKeys::date = { "now", "date"};

const std::vector<std::string>& FunctionsView::getFuncKeyList(FunctionType type)
    {
    switch (type)
        {
        using enum FunctionType;
        case TRIG: return functionKeys.trig;
        case ARITHM: return functionKeys.arithm;
        case DATE: return functionKeys.all;
        case ALL: return functionKeys.all;
        default: return  functionKeys.all; //should not happen.
        }
    }
;void FunctionsView::addFunctions(FunctionType type)
    {
    for(auto key: getFuncKeyList(type))
        defs.insert(key);
    }

void FunctionsView::removeFunctions(FunctionType type)
    {
    for(auto key: getFuncKeyList(type))
        defs.erase(key);
    }

FunctionsView::FunctionsView(Globals& globals) 
    : globals(globals) 
    {
    for (auto& func : globals.functionDefs.functions)
        {
        defs.insert(func.first);
        }
    }

FunctionDef* FunctionsView::get(const std::string& key)
    {
    if(exists(key))
        return globals.functionDefs.functions[key];
    else
        return nullptr;
    }

FunctionKeys FunctionsView::functionKeys;

FunctionKeys::FunctionKeys()
    {
    all.insert(all.end(), trig.begin(), trig.end());
    all.insert(all.end(), arithm.begin(), arithm.end());
    all.insert(all.end(), date.begin(), date.end());
    }
